pennylane/gradients/general_shift_rules.py

```diff
@@ -13,69 +13,109 @@
 # limitations under the License.
 """Contains a function for generating generalized parameter shift rules."""
 import functools
+import itertools
 import warnings
+
 import numpy as np
 import pennylane as qml
 
 
-@functools.lru_cache(maxsize=None)
-def get_shift_rule(frequencies, shifts=None):
-    r"""Computes the parameter shift rule for a unitary based on its generator's eigenvalue frequency
-    spectrum.
+def _process_gradient_recipe(gradient_recipe, tol=1e-10):
+    """Utility function to process gradient recipes.
 
-    To compute gradients of circuit parameters in variational quantum algorithms, expressions for
-    cost function first derivatives with respect to the variational parameters can be cast into
-    linear combinations of expectation values at shifted parameter values. These "gradient recipes"
-    can be obtained from the unitary generator's eigenvalue frequency spectrum. Details can be
-    found in `arXiv:2107.12390 <https://arxiv.org/abs/2107.12390>`__.
+    This utility function accepts coefficients and shift values, and performs the following
+    processing:
+
+    - Removes all small (within absolute tolerance ``tol``) coefficients and shifts
+
+    - Removes terms with the coefficients are 0
+
+    - Terms with the same shift value are combined into a single term.
+
+    - Finally, the terms are sorted according to the absolute value of ``shift``,
+      ensuring that, if there is a zero-shift term, this is returned first.
+    """
+    # remove all small coefficients and shifts
+    gradient_recipe[np.abs(gradient_recipe) < tol] = 0
+
+    # remove columns where the coefficients are 0
+    gradient_recipe = gradient_recipe[:, ~(gradient_recipe[0] == 0)]
+
+    # sort columns according to abs(shift)
+    gradient_recipe = gradient_recipe[:, np.argsort(np.abs(gradient_recipe)[-1])]
+    round_tol = int(-np.log10(tol))
+    unique_shifts = np.unique(np.round(gradient_recipe[-1], round_tol))
+
+    if gradient_recipe.shape[-1] != len(unique_shifts):
+        # sum columns that have the same shift value
+        gradient_recipe = np.stack(
+            [
+                np.stack(
+                    [
+                        np.sum(
+                            gradient_recipe[
+                                :, np.nonzero(np.round(gradient_recipe[-1], round_tol) == b)[0]
+                            ],
+                            axis=1,
+                        )[0]
+                        for b in unique_shifts
+                    ]
+                ),
+                unique_shifts,
+            ]
+        )
+
+        # sort columns according to abs(shift)
+        gradient_recipe = gradient_recipe[:, np.argsort(np.abs(gradient_recipe)[-1])]
+
+    return gradient_recipe
+
+
+@functools.lru_cache(maxsize=None)
+def eigvals_to_frequencies(eigvals):
+    r"""Convert an eigenvalue spectra to frequency values, defined
+    as the the set of positive, unique, differences of the spectra.

```

Comment:
```suggestion
    r"""Convert an eigenvalue spectrum to frequency values, defined
    as the the set of positive, unique differences of the eigenvalues in the spectrum.
```

Could even avoid the word spectrum altogether....

Reply:
Like above, could be useful if the data structure you call "eigenvalue spectrum" is defined here right in the first sentence?